### PR TITLE
oneDAL error handling improvements 

### PR DESCRIPTION
**Pull request by mchernov-intel**
eh

### CODE REVIEW

File: source/elements/oneDAL/source/common_interface/error_handling.rst [hunk 1]
```diff
@@ -14,8 +14,8 @@ Exception classification
 Exception classification in oneDAL is aligned with C++ Standard Library
 classification. oneDAL shall introduce abstract classes that define the base
 class in the hierarchy of exception classes. Concrete exception classes are
-derived from respective C++ Standard Library exception classes. oneDAL library
-shall throw exceptions represented with concrete classes.
+derived from the respective C++ Standard Library exception classes. oneDAL
+library shall throw exceptions represented with concrete classes.
```
Comment (outoftardis) [inline source/elements/oneDAL/source/common_interface/error_handling.rst:18]: oneDAL already has "library" in its name, so remove "library" here

File: source/elements/oneDAL/source/common_interface/error_handling.rst [hunk 2]
```diff
@@ -137,3 +139,28 @@ reported via ``oneapi::dal::internal_error``.
    * - ``oneapi::dal::internal_error``
      - Reports all runtime errors that couldn't be assigned to other inheritors.
      - No
+
+All memory allocation errors are represented by ``oneapi::dal::bad_alloc``. They
+shall be divided into 2 groups depending on where the error occured:
```
Comment (outoftardis) [inline source/elements/oneDAL/source/common_interface/error_handling.rst:144]:
```suggestion
shall be divided into two groups depending on where the error occurred:
```

File: source/elements/oneDAL/source/common_interface/error_handling.rst [hunk 3]
```diff
@@ -137,3 +139,28 @@ reported via ``oneapi::dal::internal_error``.
...
+
+- Host memory allocation error.
+- Device memory allocation error.
```
Comment (outoftardis) [inline source/elements/oneDAL/source/common_interface/error_handling.rst:147]:
```suggestion
- Host memory allocation error
- Device memory allocation error
```

File: source/elements/oneDAL/source/common_interface/error_handling.rst [hunk 4]
```diff
@@ -137,3 +139,28 @@ reported via ``oneapi::dal::internal_error``.
...
+
+.. list-table::
+   :widths: 30 65 5
+   :header-rows: 1
+
+   * - Exception
+     - Description
+     - Abstract
+   * - ``oneapi::dal::host_bad_alloc``
+     - Reports failures to allocate storage on the host.
```
Comment (andreynv) [inline source/elements/oneDAL/source/common_interface/error_handling.rst:162]: Report failure to allocate?

File: source/elements/oneDAL/source/common_interface/error_handling.rst (hunk 1 above)
Comment (mchernov-intel) [reply source/elements/oneDAL/source/common_interface/error_handling.rst]: fixed

File: source/elements/oneDAL/source/common_interface/error_handling.rst (hunk 2 above)
Comment (mchernov-intel) [reply source/elements/oneDAL/source/common_interface/error_handling.rst]: fixed

File: source/elements/oneDAL/source/common_interface/error_handling.rst (hunk 4 above)
Comment (mchernov-intel) [reply source/elements/oneDAL/source/common_interface/error_handling.rst]: fixed

**Comment by mchernov-intel** [comment]
@andreynv @outoftardis all comments were fixed

File: source/elements/oneDAL/source/common_interface/error_handling.rst (hunk 3 above)
Comment (mchernov-intel) [reply source/elements/oneDAL/source/common_interface/error_handling.rst]: fixed

File: source/elements/oneDAL/source/common_interface/error_handling.rst [hunk 5]
```diff
@@ -101,19 +101,21 @@ shall be divided into the following groups:
...
+   * - ``oneapi::dal::unsupported_device``
+     - Reports situations when a device is not supported.
      - No
 
-If an error occurs during function execution after preconditions and invariants
-were checked, it is reported via ``ondeal::runtime_error`` inheritors. oneDAL
-distinguishes errors happened during interaction with OS facilities and errors
-of destination type's range in internal computations, while other errors are
-reported via ``oneapi::dal::internal_error``.
+Errors occured during the execution of oneDAL functionality are represented with
```
Comment (outoftardis) [inline source/elements/oneDAL/source/common_interface/error_handling.rst:112]:
```suggestion
Errors that occur during the execution of oneDAL functionality are represented with
```

File: source/elements/oneDAL/source/common_interface/error_handling.rst [hunk 6]
```diff
@@ -101,19 +101,21 @@ shall be divided into the following groups:
...
+``oneapi::dal::runtime_error``. Two main groups of errors are distinguished:
+
+- destination type range errors
+- OS facilities interaction errors
```
Comment (outoftardis) [inline source/elements/oneDAL/source/common_interface/error_handling.rst:116]: is there a way to name them differently? 4 nouns in a row is a bit too much

File: source/elements/oneDAL/source/common_interface/error_handling.rst [hunk 7]
```diff
@@ -101,19 +101,21 @@ shall be divided into the following groups:
...
+     - Reports situations when a device is not supported.
      - No
 
-If an error occurs during function execution after preconditions and invariants
-were checked, it is reported via ``ondeal::runtime_error`` inheritors. oneDAL
-distinguishes errors happened during interaction with OS facilities and errors
-of destination type's range in internal computations, while other errors are
-reported via ``oneapi::dal::internal_error``.
+Errors occured during the execution of oneDAL functionality are represented with
+``oneapi::dal::runtime_error``. Two main groups of errors are distinguished:
```
Comment (outoftardis) [inline source/elements/oneDAL/source/common_interface/error_handling.rst:113]: shouldn't we say "shall be distinguished"?

File: source/elements/oneDAL/source/common_interface/error_handling.rst [hunk 8]
```diff
@@ -137,3 +139,28 @@ reported via ``oneapi::dal::internal_error``.
    * - ``oneapi::dal::internal_error``
      - Reports all runtime errors that couldn't be assigned to other inheritors.
      - No
+
+All memory allocation errors are represented by ``oneapi::dal::bad_alloc``. They
+shall be divided into two groups depending on where the error occurred:
```
Comment (outoftardis) [inline source/elements/oneDAL/source/common_interface/error_handling.rst:144]: this is probably better
```suggestion
All memory allocation errors are represented by ``oneapi::dal::bad_alloc``. They
shall be divided into two groups based on where they occur:
```

Comment (mchernov-intel) [reply source/elements/oneDAL/source/common_interface/error_handling.rst]: fixed

File: source/elements/oneDAL/source/common_interface/error_handling.rst (hunk 7 above)
Comment (mchernov-intel) [reply source/elements/oneDAL/source/common_interface/error_handling.rst]: fixed

File: source/elements/oneDAL/source/common_interface/error_handling.rst (hunk 6 above)
Comment (mchernov-intel) [reply source/elements/oneDAL/source/common_interface/error_handling.rst]: fixed

File: source/elements/oneDAL/source/common_interface/error_handling.rst (hunk 5 above)
Comment (mchernov-intel) [reply source/elements/oneDAL/source/common_interface/error_handling.rst]: fixed

**Comment by mchernov-intel** [comment]
@outoftardis the latest portion of comments was fixed

**Comment by andreynv** [comment]
do we have any other comments after previous release to fix?

**Comment by mchernov-intel** [comment]
@andreynv I have already fixed comments after previous release in the first commit in this pull request. So, no.

**Comment by outoftardis** [comment]
looks good to me

**Comment by andreynv** [comment]
discussed offline - couple of places should be updated

**Comment by mchernov-intel** [comment]
@andreynv fixed

**Review by andreynv — APPROVED** [review]

**Pull request merged**